src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomerServiceImplIT.java

```diff
@@ -0,0 +1,463 @@
+    private CustomerService customerService;
+    private Customer customer;
+
+
+    private List<String> errorCallBackMessages;
+    private List<String> warningCallBackMessages;
+    private List<Throwable> errorCallBackExceptions;
+
+    /**
+     * Deletes Customers from target CTP projects, then it populates target CTP project with customer test data.
+     */
+    @BeforeEach
+    void setupTest() {
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        warningCallBackMessages = new ArrayList<>();
+
+        deleteCustomers(CTP_TARGET_CLIENT);
+
+        final CustomerSyncOptions customerSyncOptions =
+                CustomerSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+                        .errorCallback(
+                            (exception, oldResource, newResource, updateActions) -> {
+                                errorCallBackMessages.add(exception.getMessage());
+                                errorCallBackExceptions.add(exception.getCause());
+                            })
+                        .warningCallback(
+                            (exception, oldResource, newResource) -> warningCallBackMessages
+                                        .add(exception.getMessage()))
+                        .build();
+
+        // Create a mock new customer in the target project.
+        CustomerDraft customerDraft = CustomerDraftBuilder
+                .of("mail@mail.com", "password")
+                .key(EXISTING_CUSTOMER_KEY)
+                .build();
+        customer = CTP_TARGET_CLIENT.execute(CustomerCreateCommand.of(customerDraft))
+                .toCompletableFuture().join().getCustomer();
+
+        customerService = new CustomerServiceImpl(customerSyncOptions);
+    }
+
+    /**
+     * Cleans up the target test data that were built in this test class.
+     */
+    @AfterAll
+    static void tearDown() {
+        deleteCustomers(CTP_TARGET_CLIENT);
+    }
+
+    @Test
+    void fetchCachedCustomerId_WithNonExistingCustomer_ShouldNotFetchACustomerId() {
+        final Optional<String> customerId = customerService.fetchCachedCustomerId("non-existing-customer-key")
+                .toCompletableFuture()
+                .join();
+        assertThat(customerId).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchCachedCustomerId_WithExistingNotCachedCustomer_ShouldFetchACustomerId() {
+        final Optional<String> customerId = customerService.fetchCachedCustomerId(EXISTING_CUSTOMER_KEY)
+                .toCompletableFuture()
+                .join();
+        assertThat(customerId).isNotEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchCachedCustomerId_WithCachedExistingCustomer_ShouldFetchFromCache() {
```

Comment:
I think some of these tests should rather be unit tests. Integration tests should be only minimum amount of tests as they're slow. In these tests you always need to create before the tests start a new customer and then clean up, and some tests don't need that (like `fetchCachedCustomerId_WithUnexpectedException_ShouldFail`). Also with this test `fetchCachedCustomerId_WithCachedExistingCustomer_ShouldFetchFromCache` you don't know if it really fetches from cache or not. Additionally in the last sprint we just cleaned up some integration tests as they were too slow, so it's good not to add too many of them again.

So I recommend to move most of the tests here to unit test and mock the request/response and leave here only a few tests to ensure that the functionality works.